hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/SQLServerParameters.java

```diff
@@ -34,24 +36,48 @@ public String process(String sql, int parameterCount) {
      */
     @Override
     public String processLimit(String sql, Object[] parameterArray, boolean hasOffset) {
-        if (isProcessingNotRequired(sql)) {
+        if ( isProcessingNotRequired( sql ) ) {
             return sql;
         }
 
         // Replace 'offset ? fetch next ? rows only' with the @P style parameters for Sql Server
         int index = hasOffset ? parameterArray.length - 1 : parameterArray.length;
-        int pos = sql.indexOf( " offset ?" );
+        // See SQLServer2012LimitHandler#getOffsetFetch
+        String offsetQueryString;
+        int pos = sql.indexOf( " offset ? rows fetch next ? rows only" );
+        if ( pos > -1 ) {
+            offsetQueryString = " offset @P" + index++;
+        }
+        else {
+            pos = sql.indexOf( " offset 0 rows fetch next ? rows only" );
+            offsetQueryString = " offset 0";
+        }
         if ( pos > -1 ) {
-            String sqlProcessed = sql.substring( 0, pos ) + " offset @P" + index++ + " rows";
-            if ( sql.indexOf( " fetch next ?" ) > -1 ) {
+            String sqlProcessed = sql.substring( 0, pos ) + offsetQueryString + " rows";
+            if ( sql.contains( " fetch next ?" ) ) {
                 sqlProcessed += " fetch next @P" + index + " rows only ";
             }
             return sqlProcessed;
         }
-
+        // Different Hibernate ORM versions may use different lowecase/uppercase letters
+        if ( sql.toLowerCase( Locale.ROOT ).startsWith( "select top(?)" ) ) {
+            // 13 is the length og the string "select top(?)"
+            String sqlProcessed = "select top(@P" + index + ")" + sql.substring( 13 );
+            shiftValues( parameterArray );
+            return sqlProcessed;
+        }
```

Comment:
It's unfortunate that ORM sometimes generates such ugly SQL. This limit handler stuff is such a mess in 5. I almost feel like back porting the cleanups I did for H6. :-/

Reply:
It would be nice, but I would wait and see if other more complicated use cases pop up before refactoring 5.
Unless is something relatively easy to do.